File: src/main/java/trivia/Game.java
package trivia;

import java.util.ArrayList;
import java.util.LinkedList;

// REFACTOR ME
public class Game implements IGame {
   private static final int BOARD_SIZE = 12;
   private static final int WINNING_COINS = 6;
   private static final int INITIAL_QUESTIONS = 50;

   ArrayList players = new ArrayList();
   int[] playerPositions = new int[6];
   int[] playerCoins = new int[6];
   boolean[] playerInPenaltyBox = new boolean[6];

   LinkedList popQuestions = new LinkedList();
   LinkedList scienceQuestions = new LinkedList();
   LinkedList sportsQuestions = new LinkedList();
   LinkedList rockQuestions = new LinkedList();

   int currentPlayer = 0;
   boolean isGettingOutOfPenaltyBox;

   public Game() {
      for (int i = 0; i < INITIAL_QUESTIONS; i++) {
         popQuestions.addLast("Pop Question " + i);
         scienceQuestions.addLast(("Science Question " + i));
         sportsQuestions.addLast(("Sports Question " + i));
         rockQuestions.addLast(createRockQuestion(i));
      }
   }

   public String createRockQuestion(int index) {
      return "Rock Question " + index;
   }

   public boolean hasEnoughPlayers() {
      return (howManyPlayers() >= 2);
   }

   public boolean add(String playerName) {
      playerPositions[howManyPlayers()] = 1;
      playerCoins[howManyPlayers()] = 0;
      playerInPenaltyBox[howManyPlayers()] = false;
      players.add(playerName);

      System.out.println(playerName + " was added");
      System.out.println("They are player number " + players.size());
      return true;
   }

   public int howManyPlayers() {
      return players.size();
   }

   public void roll(int roll) {
      System.out.println(players.get(currentPlayer) + " is the current player");
      System.out.println("They have rolled a " + roll);

      if (playerInPenaltyBox[currentPlayer]) {
         if (roll % 2 != 0) {
            isGettingOutOfPenaltyBox = true;

            System.out.println(players.get(currentPlayer) + " is getting out of the penalty box");
            movePlayer(roll);

            System.out.println(players.get(currentPlayer)
                               + "'s new location is "
                               + playerPositions[currentPlayer]);
            System.out.println("The category is " + currentCategory());
            askQuestion();
         } else {
            System.out.println(players.get(currentPlayer) + " is not getting out of the penalty box");
            isGettingOutOfPenaltyBox = false;
         }

      } else {

         movePlayer(roll);

         System.out.println(players.get(currentPlayer)
                            + "'s new location is "
                            + playerPositions[currentPlayer]);
         System.out.println("The category is " + currentCategory());
         askQuestion();
      }

   }

   private void movePlayer(int roll) {
      playerPositions[currentPlayer] = playerPositions[currentPlayer] + roll;
      if (playerPositions[currentPlayer] > BOARD_SIZE) playerPositions[currentPlayer] = playerPositions[currentPlayer] - BOARD_SIZE;
   }

   private void askQuestion() {
      if (currentCategory() == "Pop")
         System.out.println(popQuestions.removeFirst());
      if (currentCategory() == "Science")
         System.out.println(scienceQuestions.removeFirst());
      if (currentCategory() == "Sports")
         System.out.println(sportsQuestions.removeFirst());
      if (currentCategory() == "Rock")
         System.out.println(rockQuestions.removeFirst());
   }


   private String currentCategory() {
      if (playerPositions[currentPlayer] - 1 == 0) return "Pop";
      if (playerPositions[currentPlayer] - 1 == 4) return "Pop";
      if (playerPositions[currentPlayer] - 1 == 8) return "Pop";
      if (playerPositions[currentPlayer] - 1 == 1) return "Science";
      if (playerPositions[currentPlayer] - 1 == 5) return "Science";
      if (playerPositions[currentPlayer] - 1 == 9) return "Science";
      if (playerPositions[currentPlayer] - 1 == 2) return "Sports";
      if (playerPositions[currentPlayer] - 1 == 6) return "Sports";
      if (playerPositions[currentPlayer] - 1 == 10) return "Sports";
      return "Rock";
   }

   public boolean handleCorrectAnswer() {
      if (playerInPenaltyBox[currentPlayer]) {
         if (isGettingOutOfPenaltyBox) {
            System.out.println("Answer was correct!!!!");
            playerCoins[currentPlayer]++;
            System.out.println(players.get(currentPlayer)
                               + " now has "
                               + playerCoins[currentPlayer]
                               + " Gold Coins.");

            boolean winner = isGameStillInProgress();
            nextPlayer();

            return winner;
         } else {
            nextPlayer();
            return true;
         }


      } else {

         System.out.println("Answer was corrent!!!!");
         playerCoins[currentPlayer]++;
         System.out.println(players.get(currentPlayer)
                            + " now has "
                            + playerCoins[currentPlayer]
                            + " Gold Coins.");

         boolean winner = isGameStillInProgress();
         nextPlayer();

         return winner;
      }
   }

   private void nextPlayer() {
      currentPlayer++;
      if (currentPlayer == players.size()) currentPlayer = 0;
   }

   public boolean wrongAnswer() {
      System.out.println("Question was incorrectly answered");
      System.out.println(players.get(currentPlayer) + " was sent to the penalty box");
      playerInPenaltyBox[currentPlayer] = true;

      nextPlayer();
      return true;
   }


   private boolean isGameStillInProgress() {
      return !(playerCoins[currentPlayer] == WINNING_COINS);
   }
}

File: python/game.py
from .i_game import IGame
from .player import Player
from .question_deck import QuestionDeck

class Game(IGame):
    BOARD_SIZE = 12
    WINNING_COINS = 6
    INITIAL_QUESTIONS = 50

    def __init__(self):
        self.players = []
        self.question_deck = QuestionDeck(self.INITIAL_QUESTIONS)
        self.current_player_idx = 0
        self.is_getting_out_of_penalty_box = False

    def add(self, player_name: str) -> bool:
        player = Player(player_name)
        self.players.append(player)
        print(f"{player_name} was added")
        print(f"They are player number {len(self.players)}")
        return True

    def how_many_players(self) -> int:
        return len(self.players)

    def roll(self, roll: int):
        player = self.players[self.current_player_idx]
        print(f"{player.name} is the current player")
        print(f"They have rolled a {roll}")

        if player.in_penalty_box:
            if roll % 2 != 0:
                self.is_getting_out_of_penalty_box = True
                print(f"{player.name} is getting out of the penalty box")
                self._execute_turn(player, roll)
            else:
                print(f"{player.name} is not getting out of the penalty box")
                self.is_getting_out_of_penalty_box = False
        else:
            self._execute_turn(player, roll)

    def _execute_turn(self, player: Player, roll: int):
        player.advance_by(roll, self.BOARD_SIZE)
        print(f"{player.name}'s new location is {player.position}")
        print(f"The category is {self._current_category()}")
        self._ask_question()

    def _ask_question(self):
        category = self._current_category()
        print(self.question_deck.next_question(category))

    def _current_category(self) -> str:
        player = self.players[self.current_player_idx]
        category_map = {0: "Pop", 1: "Science", 2: "Sports"}
        return category_map.get((player.position - 1) % 4, "Rock")

    def handle_correct_answer(self) -> bool:
        player = self.players[self.current_player_idx]
        if player.in_penalty_box:
            if self.is_getting_out_of_penalty_box:
                print("Answer was correct!!!!")
                player.in_penalty_box = False
                player.add_coin()
                print(f"{player.name} now has {player.coins} Gold Coins.")
                winner_in_progress = self._is_game_still_in_progress()
                self._next_player()
                return winner_in_progress
            else:
                self._next_player()
                return True
        else:
            print("Answer was correct!!!!")
            player.add_coin()
            print(f"{player.name} now has {player.coins} Gold Coins.")
            winner_in_progress = self._is_game_still_in_progress()
            self._next_player()
            return winner_in_progress

    def _next_player(self):
        self.current_player_idx += 1
        if self.current_player_idx == len(self.players):
            self.current_player_idx = 0

    def wrong_answer(self) -> bool:
        player = self.players[self.current_player_idx]
        print("Question was incorrectly answered")
        print(f"{player.name} was sent to the penalty box")
        player.in_penalty_box = True
        self._next_player()
        return True

    def _is_game_still_in_progress(self) -> bool:
        return not self.players[self.current_player_idx].has_won(self.WINNING_COINS)
